# Import libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

# Load data

In [2]:
df = pd.read_csv("C:/Saurabh/My phone/datasets/supermarket_marketing/Market_Basket_Optimisation.csv")
df.head()

Unnamed: 0 Unnamed: 1  Unnamed: 2        Unnamed: 3    Unnamed: 4  \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

         Unnamed: 5 Unnamed: 6      Unnamed: 7    Unnamed: 8    Unnamed: 9  \
0  whole weat flour       yams  cottage cheese  energy drink  tomato juice   
1               NaN        NaN             NaN           NaN           NaN   
2               NaN        NaN             NaN           NaN           NaN   
3               NaN        NaN             NaN           NaN           NaN   
4               NaN        NaN             NaN           NaN           NaN   

   ... Unnamed: 11 Unnamed: 12 Unnamed: 13    Unnamed: 14 Unnamed: 15  \
0  ...   green tea       honey       salad  mineral water      salmon   
1  ...         NaN         NaN         NaN            NaN         NaN   
2  ...         NaN         NaN         NaN            NaN         NaN   
3  ...         NaN         NaN         NaN            NaN         NaN   
4  ...         NaN         NaN         NaN            NaN         NaN   

         Unnamed: 16      Unnamed: 17 Unnamed: 18 Unnamed: 19 Unnamed: 20  
0  antioxydant juice  frozen smoothie     spinach   olive oil         NaN  
1                NaN              NaN         NaN         NaN         NaN  
2                NaN              NaN         NaN         NaN         NaN  
3                NaN              NaN         NaN         NaN         NaN  
4                NaN              NaN         NaN         NaN         NaN  

[5 rows x 21 columns]

In [3]:
df['Unnamed: 20'].isnull().sum()

7501

In [4]:
df.shape

(7501, 21)

Since this column is totally empty, we'll drop it.

In [5]:
df = df.drop('Unnamed: 20', axis = 1)

To remove NaN values which came from the format of the data being from '.csv' format we'll have to convert the data into list and then remove those values.

In [6]:
df_list = df.values.tolist()

Since value of NaN is not equal to Nan in python, since it is null.

In [7]:
new_df = []
li = []

In [8]:
for b in range(len(df_list)):
    for a in range(len(df_list[b])):
        if df_list[b][a]==df_list[b][a]:
            li.append(df_list[b][a])
    new_df.append(li)
    li = []

In [9]:
len(new_df)

7501

In [10]:
from mlxtend.preprocessing import TransactionEncoder
te = TransactionEncoder()
te_arr = te.fit(new_df).transform(new_df)

In [11]:
data = pd.DataFrame(te_arr, columns=te.columns_)
data.head()

asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0       False     True               True      False     True        False   
1       False    False              False      False    False        False   
2       False    False              False      False    False        False   
3       False    False              False      False     True        False   
4       False    False              False      False    False        False   

   bacon  barbecue sauce  black tea  blueberries  ...  turkey  vegetables mix  \
0  False           False      False        False  ...   False            True   
1  False           False      False        False  ...   False           False   
2  False           False      False        False  ...   False           False   
3  False           False      False        False  ...    True           False   
4  False           False      False        False  ...   False           False   

   water spray  white wine  whole weat flour  whole wheat pasta  \
0        False       False              True              False   
1        False       False             False              False   
2        False       False             False              False   
3        False       False             False              False   
4        False       False             False              False   

   whole wheat rice   yams  yogurt cake  zucchini  
0             False   True        False     False  
1             False  False        False     False  
2             False  False        False     False  
3             False  False        False     False  
4              True  False        False     False  

[5 rows x 120 columns]

1. True = This item is in the transaction.
2. False = This item is not in the transaction.

In [12]:
from mlxtend.frequent_patterns import apriori
frequent_itemsets = apriori(data, min_support=0.7, use_colnames=True)
frequent_itemsets

Empty DataFrame
Columns: [support, itemsets]
Index: []

In [13]:
from mlxtend.frequent_patterns import association_rules
rules = association_rules(frequent_itemsets, metric='lift', min_threshold=1)
rules.head()

ValueError: cannot call `vectorize` on size 0 inputs unless `otypes` is set

There might be some issue with the packeges so we will have to calculate everything manually.

In [14]:
lil = []
for a in range(len(new_df)):
    for b in range(len(new_df[a])):
        lil.append(new_df[a][b])

In [15]:
len(lil)

29363

In [16]:
stripped = [s.strip() for s in lil]

In [17]:
output = []
for a in stripped:
    if a not in output:
        output.append(a)

In [18]:
len(output)

119

In [19]:
output

['shrimp',
 'almonds',
 'avocado',
 'vegetables mix',
 'green grapes',
 'whole weat flour',
 'yams',
 'cottage cheese',
 'energy drink',
 'tomato juice',
 'low fat yogurt',
 'green tea',
 'honey',
 'salad',
 'mineral water',
 'salmon',
 'antioxydant juice',
 'frozen smoothie',
 'spinach',
 'olive oil',
 'burgers',
 'meatballs',
 'eggs',
 'chutney',
 'turkey',
 'milk',
 'energy bar',
 'whole wheat rice',
 'whole wheat pasta',
 'french fries',
 'soup',
 'light cream',
 'shallot',
 'frozen vegetables',
 'spaghetti',
 'pet food',
 'cookies',
 'cooking oil',
 'champagne',
 'chocolate',
 'chicken',
 'oil',
 'fresh tuna',
 'tomatoes',
 'black tea',
 'extra dark chocolate',
 'protein bar',
 'red wine',
 'pasta',
 'pepper',
 'shampoo',
 'rice',
 'sparkling water',
 'ham',
 'body spray',
 'pancakes',
 'grated cheese',
 'white wine',
 'toothpaste',
 'parmesan cheese',
 'fresh bread',
 'ground beef',
 'escalope',
 'herb & pepper',
 'tomato sauce',
 'magazines',
 'strawberries',
 'strong cheese',
 

There are 119 unique items in the whole data.

In [20]:
freq = []
leng = []

In [21]:
for a in range(len(output)):
    for c in range(len(new_df)):
        if output[a] in new_df[c]:
            leng.append(a)
    freq.append(len(leng))
    leng = []

In [22]:
freq_df = pd.DataFrame(data = {'Items':tuple(output), 'Frequency': tuple(freq)})
freq_df

Items  Frequency
0            shrimp        536
1           almonds        153
2           avocado        250
3    vegetables mix        193
4      green grapes         68
..              ...        ...
114         bramble         14
115    burger sauce         44
116         oatmeal         33
117           cream          7
118         napkins          5

[119 rows x 2 columns]

##### Calculating Support, Confidence and Lift

In [23]:
le = []
leng1 = []
leng2 = []
supp = []
supp_cnt = []
conf = []
lif = []
length = []

In [24]:
for a in range(len(output)):
    for d in range(len(output)):
        if a!=d:
            for c in range(len(new_df)):       
                if (output[a] in new_df[c]) and (output[d] in new_df[c]):
                    le.append(a)
            leng1.append(output[a])
            leng2.append(output[d])
            length.append(len(le))
            s = len(le)/len(new_df)
            supp.append(s)
            scnt = len(le)
            supp_cnt.append(scnt)
            c = len(le)/freq_df['Frequency'][d]
            conf.append(c)
            l = c/(freq_df['Frequency'][a]/len(new_df))
            lif.append(l)
            le = []

In [25]:
print(len(le))
print(len(leng1))
print(len(leng2))
print(len(supp))
print(len(supp_cnt))
print(len(conf))
print(len(lif))

0
14042
14042
14042
14042
14042
14042


In [26]:
rul = pd.DataFrame(data = {'Antecedent': leng1, 'Consequent': leng2, 'Support_count': supp_cnt,\
                           'Support': supp,'Confidence': conf, 'Lift': lif})
rul

Antecedent        Consequent  Support_count   Support  Confidence  \
0         shrimp           almonds             17  0.002266    0.111111   
1         shrimp           avocado             23  0.003066    0.092000   
2         shrimp    vegetables mix             20  0.002666    0.103627   
3         shrimp      green grapes              8  0.001067    0.117647   
4         shrimp  whole weat flour              7  0.000933    0.100000   
...          ...               ...            ...       ...         ...   
14037    napkins           ketchup              0  0.000000    0.000000   
14038    napkins           bramble              0  0.000000    0.000000   
14039    napkins      burger sauce              0  0.000000    0.000000   
14040    napkins           oatmeal              1  0.000133    0.030303   
14041    napkins             cream              0  0.000000    0.000000   

            Lift  
0       1.554934  
1       1.287485  
2       1.450197  
3       1.646400  
4       1.399440  
...          ...  
14037   0.000000  
14038   0.000000  
14039   0.000000  
14040  45.460606  
14041   0.000000  

[14042 rows x 6 columns]

In [27]:
rul['Support_count'].describe(percentiles=[0.5,0.75,0.77,0.8,0.9,0.95,0.999])

count    14042.000000
mean        10.598205
std         25.311077
min          0.000000
50%          3.000000
75%          9.000000
77%         10.000000
80%         12.000000
90%         26.000000
95%         45.000000
99.9%      274.000000
max        448.000000
Name: Support_count, dtype: float64

In [28]:
new_rul=rul[rul['Support_count']>=10].reset_index(drop=True)
new_rul

Antecedent      Consequent  Support_count   Support  Confidence  \
0           shrimp         almonds             17  0.002266    0.111111   
1           shrimp         avocado             23  0.003066    0.092000   
2           shrimp  vegetables mix             20  0.002666    0.103627   
3           shrimp  cottage cheese             20  0.002666    0.083682   
4           shrimp    energy drink             21  0.002800    0.105000   
...            ...             ...            ...       ...         ...   
3485  burger sauce       chocolate             10  0.001333    0.008137   
3486  burger sauce         chicken             12  0.001600    0.026667   
3487  burger sauce     ground beef             11  0.001466    0.014925   
3488       oatmeal   mineral water             13  0.001733    0.007271   
3489       oatmeal       chocolate             10  0.001333    0.008137   

          Lift  
0     1.554934  
1     1.287485  
2     1.450197  
3     1.171080  
4     1.469412  
...        ...  
3485  1.387122  
3486  4.546061  
3487  2.544437  
3488  1.652651  
3489  1.849496  

[3490 rows x 6 columns]

In [29]:
new_rul['Lift'].describe(percentiles=[0.5,0.75,0.8,0.9,0.95,0.999])

count    3490.000000
mean        1.578393
std         0.730692
min         0.302173
50%         1.454614
75%         1.801260
80%         1.903716
90%         2.291162
95%         2.693962
99.9%       8.628843
max         8.890601
Name: Lift, dtype: float64

In [30]:
lift_drop = new_rul[new_rul['Lift']>=2].reset_index(drop=True)
lift_drop

Antecedent            Consequent  Support_count   Support  Confidence  \
0          shrimp           light cream             19  0.002533    0.162393   
1          shrimp     frozen vegetables            125  0.016664    0.174825   
2          shrimp                   oil             26  0.003466    0.150289   
3          shrimp              tomatoes             84  0.011199    0.163743   
4          shrimp  extra dark chocolate             14  0.001866    0.155556   
..            ...                   ...            ...       ...         ...   
577       ketchup                  milk             10  0.001333    0.010288   
578       ketchup              pancakes             12  0.001600    0.016830   
579  burger sauce             spaghetti             18  0.002400    0.013783   
580  burger sauce               chicken             12  0.001600    0.026667   
581  burger sauce           ground beef             11  0.001466    0.014925   

         Lift  
0    2.272595  
1    2.446574  
2    2.103205  
3    2.291481  
4    2.176907  
..        ...  
577  2.338509  
578  3.825577  
579  2.349610  
580  4.546061  
581  2.544437  

[582 rows x 6 columns]

In [31]:
pd.options.display.max_rows = 999 
#Using this so that we can see all the rows and check if the table makes sence.

fin = lift_drop.sort_values(by = ['Lift','Confidence'], ascending=False).reset_index(drop=True)
fin

Antecedent            Consequent  Support_count   Support  \
0    mushroom cream sauce                 pasta             20  0.002666   
1                   pasta  mushroom cream sauce             20  0.002666   
2                red wine              pet food             12  0.001600   
3                pet food              red wine             12  0.001600   
4             light cream  extra dark chocolate             12  0.001600   
5    extra dark chocolate           light cream             12  0.001600   
6              fresh tuna         fromage blanc             18  0.002400   
7           fromage blanc            fresh tuna             18  0.002400   
8                hot dogs       sparkling water             12  0.001600   
9         sparkling water              hot dogs             12  0.001600   
10                cereals     antioxydant juice             11  0.001466   
11      antioxydant juice               cereals             11  0.001466   
12         vegetables mix          green grapes             11  0.001466   
13           green grapes        vegetables mix             11  0.001466   
14          fromage blanc                 honey             25  0.003333   
15                  honey         fromage blanc             25  0.003333   
16             fresh tuna           french wine             19  0.002533   
17            french wine            fresh tuna             19  0.002533   
18                chicken           light cream             34  0.004533   
19            light cream               chicken             34  0.004533   
20            light cream                   oil             13  0.001733   
21                    oil           light cream             13  0.001733   
22               red wine                  rice             19  0.002533   
23                   rice              red wine             19  0.002533   
24               escalope                 pasta             44  0.005866   
25                  pasta              escalope             44  0.005866   
26                chicken          burger sauce             12  0.001600   
27           burger sauce               chicken             12  0.001600   
28                    oil               carrots             12  0.001600   
29                carrots                   oil             12  0.001600   
30                 shrimp                 pasta             38  0.005066   
31                  pasta                shrimp             38  0.005066   
32            cooking oil            toothpaste             14  0.001866   
33             toothpaste           cooking oil             14  0.001866   
34            french wine             black tea             10  0.001333   
35              black tea           french wine             10  0.001333   
36              olive oil     whole wheat pasta             60  0.007999   
37      whole wheat pasta             olive oil             60  0.007999   
38                    oil          tomato sauce             10  0.001333   
39           tomato sauce                   oil             10  0.001333   
40                carrots                   ham             12  0.001600   
41                    ham               carrots             12  0.001600   
42                chicken  extra dark chocolate             21  0.002800   
43   extra dark chocolate               chicken             21  0.002800   
44         vegetables mix                  rice             14  0.001866   
45                   rice        vegetables mix             14  0.001866   
46            ground beef          tomato sauce             40  0.005333   
47           tomato sauce           ground beef             40  0.005333   
48               pancakes               ketchup             12  0.001600   
49                ketchup              pancakes             12  0.001600   
50                 salmon          green grapes             11  0.001466   
51           green grapes                salmon             11 

We will remove every alternate row so that we can have a pair with greater Confidence value, which makes more sence.

In [32]:
final_df = fin[fin.index.isin([i for i in range(940) if i%2 == 0])].reset_index(drop = True)
final_df

Antecedent            Consequent  Support_count   Support  \
0    mushroom cream sauce                 pasta             20  0.002666   
1                red wine              pet food             12  0.001600   
2             light cream  extra dark chocolate             12  0.001600   
3              fresh tuna         fromage blanc             18  0.002400   
4                hot dogs       sparkling water             12  0.001600   
5                 cereals     antioxydant juice             11  0.001466   
6          vegetables mix          green grapes             11  0.001466   
7           fromage blanc                 honey             25  0.003333   
8              fresh tuna           french wine             19  0.002533   
9                 chicken           light cream             34  0.004533   
10            light cream                   oil             13  0.001733   
11               red wine                  rice             19  0.002533   
12               escalope                 pasta             44  0.005866   
13                chicken          burger sauce             12  0.001600   
14                    oil               carrots             12  0.001600   
15                 shrimp                 pasta             38  0.005066   
16            cooking oil            toothpaste             14  0.001866   
17            french wine             black tea             10  0.001333   
18              olive oil     whole wheat pasta             60  0.007999   
19                    oil          tomato sauce             10  0.001333   
20                carrots                   ham             12  0.001600   
21                chicken  extra dark chocolate             21  0.002800   
22         vegetables mix                  rice             14  0.001866   
23            ground beef          tomato sauce             40  0.005333   
24               pancakes               ketchup             12  0.001600   
25                 salmon          green grapes             11  0.001466   
26   mushroom cream sauce              escalope             43  0.005733   
27             fresh tuna                 honey             30  0.003999   
28                 turkey        barbecue sauce             19  0.002533   
29        frozen smoothie       sparkling water             11  0.001466   
30        parmesan cheese     whole wheat pasta             16  0.002133   
31      frozen vegetables                   tea             10  0.001333   
32            cooking oil                 bacon             12  0.001600   
33             fresh tuna               almonds             12  0.001600   
34            light cream           cooking oil             21  0.002800   
35          herb & pepper          tomato sauce             18  0.002400   
36          grated cheese           nonfat milk             14  0.001866   
37              olive oil  extra dark chocolate             20  0.002666   
38               red wine          tomato sauce             10  0.001333   
39                 melons               avocado             10  0.001333   
40             fresh tuna       parmesan cheese             11  0.001466   
41                  honey      whole weat flour             11  0.001466   
42            ground beef         herb & pepper            120  0.015998   
43                 turkey              pet food             10  0.001333   
44           tomato juice           fresh bread             32  0.004266   
45               pancakes       gluten free bar             16  0.002133   
46               brownies        cottage cheese             26  0.003466   
47                   soup            body spray             14  0.001866   
48                cookies               shallot             15  0.002000   
49                   soup           light cream             19  0.002533   
50          herb & pepper   clothes accessories             10  0.001333   
51          herb & pepper                  rice             22 

In [33]:
final_df[['Antecedent', 'Consequent']]

Antecedent            Consequent
0    mushroom cream sauce                 pasta
1                red wine              pet food
2             light cream  extra dark chocolate
3              fresh tuna         fromage blanc
4                hot dogs       sparkling water
5                 cereals     antioxydant juice
6          vegetables mix          green grapes
7           fromage blanc                 honey
8              fresh tuna           french wine
9                 chicken           light cream
10            light cream                   oil
11               red wine                  rice
12               escalope                 pasta
13                chicken          burger sauce
14                    oil               carrots
15                 shrimp                 pasta
16            cooking oil            toothpaste
17            french wine             black tea
18              olive oil     whole wheat pasta
19                    oil          tomato sauce
20                carrots                   ham
21                chicken  extra dark chocolate
22         vegetables mix                  rice
23            ground beef          tomato sauce
24               pancakes               ketchup
25                 salmon          green grapes
26   mushroom cream sauce              escalope
27             fresh tuna                 honey
28                 turkey        barbecue sauce
29        frozen smoothie       sparkling water
30        parmesan cheese     whole wheat pasta
31      frozen vegetables                   tea
32            cooking oil                 bacon
33             fresh tuna               almonds
34            light cream           cooking oil
35          herb & pepper          tomato sauce
36          grated cheese           nonfat milk
37              olive oil  extra dark chocolate
38               red wine          tomato sauce
39                 melons               avocado
40             fresh tuna       parmesan cheese
41                  honey      whole weat flour
42            ground beef         herb & pepper
43                 turkey              pet food
44           tomato juice           fresh bread
45               pancakes       gluten free bar
46               brownies        cottage cheese
47                   soup            body spray
48                cookies               shallot
49                   soup           light cream
50          herb & pepper   clothes accessories
51          herb & pepper                  rice
52            french wine                  rice
53              olive oil           light cream
54                chicken      whole weat flour
55                chicken             bug spray
56                 salmon               carrots
57                chicken             magazines
58               escalope               shallot
59            blueberries         grated cheese
60                    oil          tomato juice
61            cooking oil              zucchini
62              olive oil          tomato sauce
63             white wine           fresh bread
64        frozen smoothie                  gums
65               brownies          energy drink
66          herb & pepper           light cream
67         low fat yogurt               shallot
68                almonds               burgers
69            french wine               almonds
70      frozen vegetables       parmesan cheese
71            french wine           fresh bread
72                 salmon          tomato sauce
73              olive oil         strong cheese
74                   soup           blueberries
75                 salmon             magazines
76          herb & pepper             black tea
77                   soup          tomato sauce
78                   gums                salmon
79              spaghetti         strong cheese
80       whole wheat rice              eggplant
81       whole wheat rice                   ham
82                 salmon      

Final pairs according to the Market basket analysis that was performed.